File: rust/elo/src/metrics/text.rs
//! The text metric
use std::collections::HashMap;

use crate::_types::clptypes::{Message, MetricUpdate};
use crate::metrics::metrictrait::AbstractMetric;

const WEIGHT_TEXT: f32 = 0.02;

#[derive(Default, Debug)]
pub struct Text;

impl Text {
    pub fn new() -> Self {
        Self {}
    }
}

impl AbstractMetric for Text {
    fn can_parallelize(&self) -> bool {
        false
    }

    fn get_name(&self) -> String {
        String::from("text")
    }

    fn get_metric(&mut self, message: Message, _sequence_no: u32) -> MetricUpdate {
        match message {
            Message::Twitch(comment) => {
                let score = f32::max(0.0, calculate_score(comment.message.body.len()));
                self.twitch_comment_shortcut(comment, score)
            }
            Message::Discord(msg) => MetricUpdate {
                metric_name: self.get_name(),
                updates: HashMap::from([(
                    msg.author.id,
                    f32::max(0.0, calculate_score(msg.content.len())),
                )]),
            },
            _ => MetricUpdate::default(),
        }
    }
}

fn calculate_score(x: usize) -> f32 {
    -WEIGHT_TEXT * x as f32 * (x as f32 - 20.0)
}

File: python-chatdownloader/src/metrics/text.py
"""
The text metric
"""

import math
from _types import Comment

from .abstractmetric import AbstractMetric

WEIGHT_TEXT = 0.02


class Text(AbstractMetric):
    """
    Text for a metric
    """
    @classmethod
    def can_parallelize(cls) -> bool:
        return False

    @classmethod
    def get_name(cls) -> str:
        return 'text'

    @staticmethod
    def _calculate_score(x) -> float:
        return -WEIGHT_TEXT * x * (x - 20)

    def get_metric(self, comment: Comment,
                   sequence_no: int) -> dict[str, float]:
        return self._shortcut_for_this_comment_user(
            comment,
            # we mathed this
            max(0, self._calculate_score(len(comment.message.body)))
        )
